File: day15/haskell/day15.hs
import Data.Bits ((.&.))

type Generator = (Int -> Bool) -> [Int]

generate :: Int -> Int -> (Int -> Bool) -> [Int]
generate v f c = map (.&. 0xFFFF) . filter c . iterate next $ v
    where next = flip mod 2147483647 . (* f)

solve :: [Int] -> [Int] -> Int -> Int
solve a b n = length . filter id . take n $ zipWith (==) a b

part1 :: Generator -> Generator -> Int
part1 a b = solve (a $ const True) (b $ const True) 40000000

part2 :: Generator -> Generator -> Int
part2 a b = solve (a $ f 4) (b $ f 8) 5000000
    where f c = (== 0) . flip mod c

main :: IO ()
main = do
    let a = generate 883 16807
    let b = generate 879 48271
    print $ part1 a b
    print $ part2 a b

File: day15/typescript/day15.ts
import { readFile } from 'fs';
import { promisify } from 'util';

import * as R from 'ramda';

interface Generator {
    value: number;
    factor: number;
}

function* values({ value, factor }: Generator, criteria: number): IterableIterator<number> {
    for (;;) {
        if (value % criteria === 0) {
            yield value;
        }

        value = (value * factor) % 2147483647;
    }
}

const solve = (as: IterableIterator<number>, bs: IterableIterator<number>, num: number) => {
    let count = 0;
    for (const i of R.range(0, num)) {
        if ((as.next().value & 0xFFFF) === (bs.next().value & 0xFFFF)) {
            count++;
        }
    }

    return count;
};

const part1 = (a: Generator, b: Generator) => solve(
    values(a, 1),
    values(b, 1),
    40000000,
);

const part2 = (a: Generator, b: Generator) => solve(
    values(a, 4),
    values(b, 8),
    5000000,
);

(() => {
    const a = { value: 883, factor: 16807 };
    const b = { value: 879, factor: 48271 };

    console.log(part1(a, b));
    console.log(part2(a, b));
})();
